### PR TITLE
row_cache: test cache consistency during multi-partition cache updates

### PR DESCRIPTION
**Pull request by michoecho**
Adds a test reproducing https://github.com/scylladb/scylladb/issues/16759, and the instrumentation needed for it.

### CODE REVIEW

**Comment by michoecho** [comment]
I'm posting it because @tgrabiec and @avikivity asked for it in https://github.com/scylladb/scylladb/pull/17138. But I keep being skeptical about writing tests for issues which are already fixed.

File: configure.py [hunk 1]
```diff
@@ -420,7 +420,7 @@ def find_ninja():
         'description': 'a mode with optimizations and no debug checks, used for production builds',
     },
     'dev': {
-        'cxxflags': '-DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION',
+        'cxxflags': '-DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -DSCYLLA_ENABLE_PREEMPTION_SOURCE',
         'cxx_ld_flags': '',
```
Comment (avikivity) [inline configure.py:424]: > `preemption_source` is a superset of preemption_source,

of preemption_check?

Comment (michoecho) [reply configure.py:424]: Yes. Fixed.

**Review by avikivity — APPROVED** [review]

**Comment by nyh** [comment]
> I'm posting it because @tgrabiec and @avikivity asked for it in #17138. But I keep being skeptical about writing tests for issues which are already fixed.

My philosophy on why this kind of tests is still useful and important:

1. It proves to you, and to the reviewers, that you really fixed the bug and don't just **think** you fixed the bug. My (un)favorite story is about how we designed and implemented an elaborate flow-control feature for materialized views (https://www.scylladb.com/2018/12/04/worry-free-ingestion-flow-control/), and were very proud of the new feature - until a few months later we discovered that a typo caused this feature to be turned off (see its fix in da54d0fc7deceb752475770b2a03bf39b2838036)  - and nobody noticed because nobody actually had a test that failed without the fix, and the manual tests that the developer (Duarte) did were obviously run before this typo appeared in the code - and then were never run again.
2. In my experience, while developing the test you often (if given enough freedom to spend time on this, of course) think of additional edge cases that are worth testing - and you might even find loose ends in your implementation and need to fix them.
3. Some test frameworks allow you to run the test against a Scylla of arbitrary version. This allows you to check if some version of Scylla is vulnerable to this bug even if you don't know exactly what the version in question contains.
4. Collecting a large body of **functional** tests (tests which use the user-visible functionality of Scylla, not unit-test of internal implementation) makes future refactoring much safer. Imagine that a year from now, somebody will have a great idea on how to re-implement the cache or a piece of it. This and other regression tests will ensure that the re-implementation cannot break things that were already working correctly.

**Pull request merged**